sound/soc/sof/ipc3-control.c

```diff
@@ -381,31 +383,46 @@ static int sof_ipc3_bytes_ext_put(struct snd_sof_control *scontrol,
 		return -EINVAL;
 	}
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
-		return -EFAULT;
+	saved = kmemdup(cdata->data, scontrol->max_size, GFP_KERNEL);
+	if (!saved)
+		return -ENOMEM;
+
+	if (copy_from_user(cdata->data, tlvd->tlv, header.length)) {
+		ret = -EFAULT;
+		goto err_free;
+	}
 
 	if (cdata->data->magic != SOF_ABI_MAGIC) {
 		dev_err_ratelimited(scomp->dev, "Wrong ABI magic 0x%08x\n", cdata->data->magic);
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
 		dev_err_ratelimited(scomp->dev, "Incompatible ABI version 0x%08x\n",
 				    cdata->data->abi);
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	/* be->max has been verified to be >= sizeof(struct sof_abi_hdr) */
 	if (cdata->data->size > scontrol->max_size - sizeof(struct sof_abi_hdr)) {
 		dev_err_ratelimited(scomp->dev, "Mismatch in ABI data size (truncated?)\n");
-		return -EINVAL;
+		goto err_restore;
 	}
 
 	/* notify DSP of byte control updates */
-	if (pm_runtime_active(scomp->dev))
-		return sof_ipc3_set_get_kcontrol_data(scontrol, true, true);
+	if (pm_runtime_active(scomp->dev)) {
+		ret = sof_ipc3_set_get_kcontrol_data(scontrol, true, true);
+		if (ret < 0)
+			goto err_restore;
+	}
```

Comment:
saved needs to be freed even when there's no error isnt it?

Reply:
In this implementation indeed it does (and thus I probably missed it). That said I was considering something slightly different, for "saved" to be preallocated instead of being allocated every time we change the config. As in another field in the struct. That way there's no need to kmalloc/kfree it in the first place.